027-network-policy-generation-feature-gate.md

```diff
@@ -0,0 +1,63 @@
+# Network Policy Generation Feature Gate
+
+The feature would consist of a new Feature Gate which can be used to disable the generation of Network Policies by Strimzi, where the user wishes to set up a more restrictive Network Policy set.
+
+## Current situation
+
+Currently, Strimzi generates Network Policies which allow all pods in the deployment to connect to Kafka, Zookeeper and any other service that it has generated.
+
+There is no way to stop this generation.
+
+It is possible, in the select case of listeners, to further restrict the Ingress using NetworkPolicyPeers, however this does not extend to all Network Policy generation, for example Metrics.
+It is also not possible to set up Egress Network Policies. 
+
+## Motivation
+
+The current Network Policies that are generated are great for getting Strimzi up and running in a deployment where a deny-all network policy is in place.
+
+However, this does not allow users of Strimzi to reach a fully secured Network Policy set, due to Strimzi not providing the full feature set of Kubernetes Network Policies.
+
+By adding a feature gate, it allows users to disable this generation, and instead write fully secure Network Policies.
+
+This setup is valid in 2 situations:
+
+* A deployment with no deny-all Network Policy, where all pods can already communicate with each other.
+* A deployment where the consumer wishes to write their own secure Network Policies.
+
+## Proposal
+
+The proposal would be to introduce a Feature Gate which would provide the ability to turn of the generation of Network Policies by Strimzi.
+
+As the functionality already exists to generate Network Policies, a sensible naming of this gate would be `NetworkPolicyGeneration` which by default is set to `true`.
+If users require secure Network Policies, they can disable this using the regular Feature Gate syntax.
+This Feature Gate will be a long lived Feature Gate, and is not meant to phase in a feature. Instead, it should remain so users can turn off this functionality when needed.
+
+When this feature gate is disabled, all generation of Network Policies from the Strimzi Operator is also disabled.
+
+This extends to:
+
+* CruiseControl
+* KafkaCluster
+* KafkaConnectCluster
+* ZookeeperCluster
```

Comment:
What about Kafka Mirror Maker 2 and Kafka Bridge? I think it should cover all of them.

Reply:
Good point. KafkaMirrorMaker2 is already included in the PR, so I will add it here.
Kafka Bridge however does not appear to generate any Network Policies?

Reply:
Would users always want to enable/disable this feature across all these things simultaneously? I can imagine use cases for KafkaConnect where it might be necessary to do something custom because of a particular connector, but otherwise the user might be happy to have the normal Strimzi behaviour, for example.

Reply:
@scholzj As mentioned, I have added KafkaMirrorMaker2, however Kafka Bridge does not appear to have any network policy generation currently so is not included.

@tombentley This is a good point, however I would say that the most likely situation is that a user is either happy with Strimzi's generated rules, or has their own guidelines for NetworkPolicies in which they will want to write these all themselves anyway. Certainly possible but I think most users would just end up disabling all generation anyway.
I have put it in the rejected solutions, as it is a good idea if it was needed, however we can switch to multiple flags if you feel this is required.